<a href="https://colab.research.google.com/github/utkh-gt/grievance-nlp-system/blob/main/notebooks/Week2_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/Project 1/cleaned_data.csv")

df.head()

,Text,Department,clean_text
0,Street Light Not Working,Electrical,street light working
1,Street Light Not Working,Electrical,street light working
2,Garbage dump,Solid Waste (Garbage) Related,garbage dump
3,Street Light Not Working,Electrical,street light working
4,Garbage dump,Solid Waste (Garbage) Related,garbage dump


In [2]:
df.columns

Index(['Text', 'Department', 'clean_text'], dtype='object')

## **Applying TF-IDF Vectorization**

In [3]:
from sklearn.feature_extraction.text import TfidfVectorizer

X = df['clean_text']
y = df['Department']

vectorizer = TfidfVectorizer()
X_tfidf = vectorizer.fit_transform(X)

## **Training logistic regression model**

In [4]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

X_train, X_test, y_train, y_test = train_test_split(X_tfidf, y, test_size=0.2, random_state=1)

model = LogisticRegression(max_iter=1000)

model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

## **Finding accuracy and classification report**

In [6]:
from sklearn.metrics import accuracy_score

y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.9975979523528253


In [7]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


                               precision    recall  f1-score   support

                Advertisement       1.00      1.00      1.00       144
         BBMP Election Branch       0.00      0.00      0.00         1
               CORONA COVID19       0.00      0.00      0.00         3
                  Call Center       1.00      0.54      0.70        13
     E khata / Khata services       1.00      1.00      1.00       845
                    Education       0.00      0.00      0.00         2
                   Electrical       1.00      1.00      1.00      8407
                       Estate       0.00      0.00      0.00        16
                       Forest       1.00      1.00      1.00      1245
                  Health Dept       1.00      1.00      1.00       690
               Indira Canteen       1.00      0.76      0.86        33
       Information Technology       1.00      1.00      1.00         8
                        Lakes       1.00      1.00      1.00        74
     

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [12]:
df[['Text', 'Department']].drop_duplicates().shape

(185, 2)

The dataset contains only 185 unique complaint types repeated across 120k entries. As most of the people complain about the same things which lacks uniqueness (i.e. street light not working, potholes etc.). Each complaint type strongly corresponds to a single department label, making the classification task relatively straightforward. This explains the very high model accuracy.

# **Week 2 Done**